### PR TITLE
[Core] Enable decode of context length equal to max model length

### PR DESCRIPTION
**Pull request by yannicks1**
## Purpose

On the current main branch, the stopping criterion does not allow a decode on context length `max_model_len`, the last decode happens at context length `max_model_len - 1` (one token before the actual limit). 

## Proposed change
Relax the assertion/check such that a last token can be generated and sampled when the context length exactly equals `max_model_len`.

changes needed:
- [vllm/v1/core/sched/scheduler.py](https://github.com/vllm-project/vllm/compare/main...yannicks1:vllm:enable-decode-of-max-model-len?expand=1#diff-9eeca590fd99f15621897e559dba39b3ec4e7c2c65ec3c3229711689e008b5f4)
- [vllm/v1/core/sched/utils.py](https://github.com/vllm-project/vllm/compare/main...yannicks1:vllm:enable-decode-of-max-model-len?expand=1#diff-d62daddf89a3f4abaf0f507a93ac2364c0fea3a31643c9bd2c0915662c2545fa)

This restores parity with the HuggingFace [transformers](https://github.com/huggingface/transformers) behavior (see [this](https://github.com/huggingface/transformers/pull/40775) PR) and avoids rejecting otherwise-valid generation requests.

## Test Plan
Extend the existing end-to-end context length test that compares vLLM to HuggingFace Transformers:
- Test file: tests/v1/e2e/test_context_length.py
- Test logic: build a textual prompt that tokenizes to exactly `prompt_len=max_model_len - 1`, then:
  - Run vLLM generation requesting `max_tokens=2`.
  - Run HF .generate(...) requesting `max_tokens=2`.
  - Assert both return the same number of generated tokens (equaling `max_tokens`) and the same token ids.

### Failing behavior (before this PR)

Without the change the unit test fails with:

```
============================================================================================== FAILURES ==============================================================================================
________________________________________________________________________ test_max_context_length[2047-2-JackFram/llama-160m] _________________________________________________________________________

>   assert len(vllm_output_ids) == len(hf_output_ids) == max_tokens
      ^^^^^^^^^^^^^^^^^
E   assert 1 == 2
E    +  where 1 = len([15898])
E    +  and   2 = len([15898, 11587])

tests/v1/e2e/test_context_length.py:98: AssertionError
```

### Test result (after this PR)

both tests pass:
```
=========================================================================== 2 passed in 36.51s ===========================================================================
```

### CODE REVIEW

File: tests/v1/e2e/test_context_length.py [hunk 1]
```diff
@@ -4,15 +4,22 @@
...
-This test verifies the following behavior: allow a prefill that fills the
-model's maximum context length and then request a single new token.
+This test verifies the following behavior: allow prefill and decodes on the
+model's maximum context length ``max_model_len`` and get one more token.
 
 Test strategy
-- Build a textual prompt that tokenizes to exactly ``max_model_len`` tokens.
-- Run vLLM generation requesting a single new token (max_tokens=1).
-- Run HF generation on the same prompt requesting a single token too.
+- Build a textual prompt that tokenizes to exactly ``prompt_len`` tokens.
```
Comment (tdoublep) [inline tests/v1/e2e/test_context_length.py]: nit: Do you actually generate a "textual prompt"?

File: tests/v1/e2e/test_context_length.py
Comment (tdoublep) [inline tests/v1/e2e/test_context_length.py:66]: Should we also pass `max_model_len=2048` here? Just in case someone tries to run this test with a different model.

File: tests/v1/e2e/test_context_length.py (hunk 1 above)
Comment (yannicks1) [reply tests/v1/e2e/test_context_length.py]: good catch, will change that

File: vllm/v1/core/sched/scheduler.py [hunk 2]
```diff
@@ -224,7 +224,7 @@ def schedule(self) -> SchedulerOutput:
             # This is necessary when using spec decoding.
             num_new_tokens = min(
                 num_new_tokens,
-                self.max_model_len - 1 - request.num_computed_tokens)
+                self.max_model_len - request.num_computed_tokens)
```
[the remark this answers is not in the document]
Comment (yannicks1) [reply vllm/v1/core/sched/scheduler.py:227]: no, Gemini is confusing `request.num_tokens` in utils.py and `request.num_computed_tokens` in scheduler.py. 

`request.num_computed_tokens` only gets updated to `request.num_tokens` after the actual model forward and hence after the check in the scheduler, which obviously happens before the model forward call).

demonstrating this by printing `request.num_tokens` in utils.py and `request.num_computed_tokens` in scheduler.py for the (second) unit test `tests/v1/e2e/test_context_length.py::test_max_context_length`:

```
# prefill 
not entering this code block 

# first decode
in utils.py: request.num_tokens 2048
in scheduler.py: request.num_computed_tokens 2047

# second decode -> exit as stop criterion hits
in utils.py: request.num_tokens 2049
```

**Review by tdoublep — APPROVED** [review]
LGTM - thanks for fixing the corner case and adding the test to ensure it does not get broken!

**Comment by vanbasten23** [comment]
Hi @yannicks1 @tdoublep , this PR broke the lora test on TPU. If you check the CI "TPU V1 Test Part2" in the [CI for this PR](https://buildkite.com/vllm/ci/builds/33551#_) and search for the test `Running test_lora.py`. The commit before this PR doesn't fail that test ([CI](https://buildkite.com/vllm/ci/builds/33549)). Could you take a look?

**Comment by mgoin** [comment]
This seems like too much of a core change to make without explicitly consulting other committers, and I actually don’t think it makes sense. For me, `max_model_len` corresponds to the maximum context a model can attend to and this new behavior allows for a single request to have `max_model_len + 1 ` tokens.

Is there a strong need for this change? It seems arbitrary and I'd rather just stay consistent with previous behavior to avoid confusion

**Comment by tdoublep** [comment]
Let's revert until everyone is in agreement, but just to clarify:

>  max_model_len by 1 makes the kernel less efficient as it falls outside the existing tuned table.

This change *does not* lead to any kernels running on `max_model_len+1`. This change relates to doing a forward pass on `max_model_len` tokens and the "+1"-th token that you get is the token after sampling the logits at the end of that forward pass. 

>  For me, max_model_len corresponds to the maximum context a model can attend to

This is correct, but I believe the change is consistent with this definition. This extra token never makes it into any forward pass or any attention kernels.

**Comment by yannicks1** [comment]
I see that CI "TPU V1 Test Part2" was broken due to the change.
Updated GPU model runner, but did not pay attention to the TPU model runner (was a soft fail, so I didnt notice) sorry! 
I have a fix here #26348 that updates the cached tokens (we don't want to cache the additional token after sampling it. as @tpa explains correctly above it will never be part of any forward pass).

**Comment by vanbasten23** [comment]
Thanks for the response!

Here is how vLLM define the max_model_length today: https://github.com/vllm-project/vllm/blob/7cd95dc8a327d1366fb5a7a9425ef995c3f71dbd/vllm/config/model.py#L181-L183.

I feel this PR changes the definition in that now the max_model_length contains only the prompt length, not the output length. That's why it confused me. If you reach an agreement and decide this PR is the intended behavior, then I think the above definition should also be updated.
